dora/core/server/worker/src/main/java/alluxio/worker/netty/PacketReadTaskStateMachine.java

```diff
@@ -0,0 +1,489 @@
+    mStateMachine.setTrace(new DebugLoggingTracer<>(LOG));
+    mStateMachine.fireInitialTransition();
+    mContext = context;
+    mRequest = context.getRequest();
+    mChannel = channel;
+    mPacketReader = packetReader;
+  }
+
+  /**
+   * Generates a diagram describing the state transition in a .dot file.
+   * Only used for testing purposes.
+   *
+   * @param outputFile path to the output file
+   * @throws IOException when writing file fails
+   */
+  @VisibleForTesting
+  public void generateStateDiagram(Path outputFile) throws IOException {
+    try (OutputStream outFile = Files.newOutputStream(outputFile,
+        StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
+      mStateMachine.configuration().generateDotFileInto(outFile, /* printLabels */ true);
+    }
+  }
+
+  /**
+   * Start the state machine.
+   */
+  public void run() {
+    Preconditions.checkState(mStateMachine.isInState(State.CREATED),
+        "state machine cannot be restarted: expected initial state %s, "
+            + "encountered %s", State.CREATED, mStateMachine.getState());
+    fireNext(TriggerEvent.START);
+    try {
+      for (Runnable trigger = mNextTriggerEvent.getAndSet(null);
+           trigger != null;
+           trigger = mNextTriggerEvent.getAndSet(null)) {
+        trigger.run();
+      }
+    } catch (RuntimeException e) {
+      LOG.error("Unexpected exception during execution, state: {}", mStateMachine.getState(), e);
+      // close and release the channel in case there is a programming error in the state machine
+      if (mChannel != null) {
+        CommonUtils.closeChannel(mChannel);
+      }
+      throw e;
+    }
+
+    Preconditions.checkState(mStateMachine.isInState(State.COMPLETED),
+        "execution of state machine has stopped but it is not in a terminated state");
+  }
+
+  private void onReadingData() {
+    long start = mContext.getPosToQueue();
+    int packetSize = (int) Math
+        .min(mRequest.getEnd() - mContext.getPosToQueue(), mRequest.getPacketSize());
+
+    // packetSize should always be >= 0 here when reaches here.
+    Preconditions.checkState(packetSize >= 0);
+
+    DataBuffer packet = null;
+    try {
+      packet = getDataBuffer(mContext, mChannel, start, packetSize);
+    } catch (Exception e) {
+      LOG.error("Failed to read data.", e);
+      fireNext(mTriggerEventsWithParam.mFailToReadEvent,
+          new Error(AlluxioStatusException.fromThrowable(e), true));
+    }
+
+    // Put an object into the flow control queue. If the queue is full, it indicates that there are
+    // too many  in-flight packets, and we need to wait some time.
+    LOG.debug("Putting an object into the flow control queue.");
+    SendDataEvent sendDataEvent = new SendDataEvent(start, packetSize, packet);
+    mFlowControlQueue.offer(sendDataEvent);
+    LOG.debug("An object has been put into the flow control queue successfully.");
+
+    fireNext(mTriggerEventsWithParam.mSendDataEvent, sendDataEvent);
```

Comment:
Suppose you are firing a send data event here, which is simply setting `mNextTriggerEvent` and wait for the state machine to poll that runnable and run it. If a Netty thread delivers a `mFailToSendDataEvent` at this instant, it will override the runnable inside `mNextTriggerEvent`, so the state machine will transition to a different state. The bad thing about this race condition is that, `SendDataEvent` contains a `DataBuffer`, which needs expliciy releasin, otherwise it causes a memory leak. So if the event gets overriden, the buffer would be leaked.